### DISCUSSION TITLE
4.0.0.0

### DISCUSSION THREAD

**Discussion by danielkerr** [opencart/opencart#10872] [Announcements]
Its finally here! OpenCart 4.0.0.0. After many years hard work.

ADDED
PHP 8.1+ Compatibly 
New event based extension system
Startup page
CRON job page
GDPR
Product variants. You can now choose to use  create a new product variant based off a master product and link the stock quantities.
Cookie accept policy.
Admin notification system. OpenCart can now send you updates and news information directly to your OpenCart admin. 
Autoloader class
Telephone field now optional
Additional admin security, renaming admin folder, install directory, moving storage folder.
 
UPDATED
Boootstrap 5
jQuery v3.6.0 
Font Awsome 6
CKEditor 5
Renamed order_recurring to subscription system.
Sliding buttons instead of checkboxes
All forms are now using AJAX.
New installer
Upgrade system can be done fully via the admin.
Reports system.
Admin alerts.
Backup system now backs up the database in parts to avoid crashing with large tables. 
Order editor has been completely reworked so API session is not required. 

REMOVED
OCMOD
All 3rd party extensions. There is a system to download and add core extensions directly to the extension pages. 

<hr /><em>This discussion was created from the release <a href='https://github.com/opencart/opencart/releases/tag/'>4.0.0.0</a>.</em>

**Comment by PeterStudioWeb** [comment]
Cool)))

**Comment by stalker780** [comment]
CKEditor 5?

Thanks God it's still CKEditor 4 :)

**Comment by cgcmart** [comment]
ckeditor, jquery-ui, payment_method - credit_card need update.

**Comment by danielkerr** [comment]
I don't use jQuery ui

**Reply by condor2** [reply]
You are funny. 
Then delete it

**Comment by Jusufas** [comment]
What is the purpose of using product variants? Looks like it's just the same as duplicate a product. I thought it would be something like a combination of product options choices (which has long been on other platforms) when, for example, you sell clothes and a certain colour + size combination has separate codes, price, quantity.

**Reply by felipoantonoff** [reply]
It serves to avoid stock problems as a product with Color and Size, thus maintaining the use of simple options and variations for products with these cases or that want their own photo and more details aside.

Here is a good Variations system for Opencart 3: https://www.codemarket.com.br/produto/sistema-de-variacoes-no-produto-opcoes-relacionas-tamanho-cor-opencart

**Comment by danielkerr** [comment]
I did one better then the other carts. The other carts only produce variants with options. With opencarts version you can change everything including name, description, etc..

**Reply by Jusufas** [reply]
But end user anyway on frontend will see separated products. So what is the point of this? To save some time for admins? For example if you want to edit products titles, you can do it at once. I’m not saying it’s bad, but if an item has 10 different colours and 5 different sizes, and each combination has separate codes, it’s already 50 items. On admin side it is OK. But maybe in front end products should be joined into one product?

**Reply by aris9ian** [reply]
It is not good to join products in terms of SEO. Also, some marketplaces require that you provide an xml feed with product variants split and not merged because their barcodes differ.

One thing though that could be implemented is to show the variants (list or image) under the price, in the product view and the product list view so that a customer can be aware that there are variants available e.g. in different color, size etc.

**Reply by felipoantonoff** [reply]
Here is a good Variations system for Opencart 3: https://www.codemarket.com.br/produto/sistema-de-variacoes-no-produto-opcoes-relacionas-tamanho-cor-opencart

Maybe it's an idea, we did it in a similar way to big companies like Nike and Netshoes, the products are related by variations, great for cases that need size and color, where the color can be the product and the size the simple option, thus solving the problem of stocking products with related options.

**Comment by karan982** [comment]
I see There are only five payment methods are available: 

Bank Transfer
Cheque / Money Order
Cash On Delivery
Credit Card
Free Checkout

What about others like ?
PayPal
Authorize.Net 
SagePay 
Skrill
Klarna 
Square
etc.

Same for the Shipping Methods. 

​Thanks,
Kind & Best Regards
Coding Inspect

**Comment by aris9ian** [comment]
`As of OpenCart 4+ OCMOD will be removed from the OpenCart source code. Only the event system will be used for modifications.`

@danielkerr how are we going to modify the templates/views via the event system? I think with the event system you can modify only the controller/model before or after an event.

**Reply by stalker780** [reply]
I would say even more:
1. Not possible to modify any SQL requests
2. Not possible to modify any system libraries
3. Modifying templates is possible. But it's a pain in...

Say hello to good old vqmod :)

OpenCart 4 VQMod Manager
https://www.opencart.com/index.php?route=marketplace/extension/info&extension_id=42736
https://github.com/stalker780/opencart-vqmod-manager

**Reply by aris9ian** [reply]
Cool! The only case it does not seem to work is when you've changed the admin path!

**Reply by aris9ian** [reply]
**The extension is super cool!**

**Reply by stalker780** [reply]
> Cool! The only case it does not seem to work is when you've changed the admin path!

Hope [qphoria](https://github.com/qphoria) updates VQMod for OC4.

If not, I will try to sort this out myself.

**Reply by konstantinosbotonakis** [reply]
vQmod for the win?!

**Reply by felipoantonoff** [reply]
Thread with suggestion to add OCMOD to Opencart 4: https://github.com/opencart/opencart/discussions/10883

**Reply by karan982** [reply]
Does anybody know view file code like we do in Ocmod ?

for example this is view code:
              `{% for store in stores %}`
                `<a href="{{ store.href }}" target="_blank" class="dropdown-item">{{ store.name }}</a>`
              `{% endfor %}`
              
with ocmod we can set search string 
`<a href="{{ store.href }}" target="_blank" class="dropdown-item">{{ store.name }}</a>`
and replace with 
`<a href="{{ store.href }}" target="_blank" class="dropdown-item store-{{ store.id }}">{{ store.name }}</a>`

Download sample ocmod.zip in this link.
https://www.codinginspect.com/test/modify_viewfile.ocmod.zip

do not install this ocmod file on opencart store. it is only demonstration purpose only.  To acheive same out with event system.

**Does anybody know how we can acheive the same event system ?**

**Comment by danielkerr** [comment]
I'm going to write some documentation, full API and examples like how create language, template extensions

**Reply by aris9ian** [reply]
Take your time @danielkerr . We appreciate it!

**Reply by karan982** [reply]
@danielkerr Good to hear from you. I appreciate it.

Is there any help I can offer from Codinginspect ?

**Reply by felipoantonoff** [reply]
Very good Daniel, if you need help translating into Brazilian Portuguese I can help

**Reply by ptsashish** [reply]
@danielkerr how we can update the cart items data on the cart page using event? Canyon please help me on this?

**Reply by neeooh** [reply]
Hi there! Have you wrote documentation on creating modules, and perhaps some sample modules? I want to create a module that auto pre-fills a text input field in the add new product form. Can't find any dev resources. Thanks in advance!

**Comment by karan982** [comment]
Issue: Admin Folder Still exists after Rename.

Step: 1. Open Admin Dashboard Page. 
![image](https://user-images.githubusercontent.com/90453683/169997713-98406212-0eb5-422a-8283-4e884c8f9032.png)

Step 2: Rename admin folder to 'manager' 
![image](https://user-images.githubusercontent.com/90453683/169998017-afc245ec-bc2d-4104-b8b4-f3033b065827.png)

**Step 3: Admin Folder Still exists after Rename.**
![image](https://user-images.githubusercontent.com/90453683/169998252-5d3117ac-de95-4adb-b86f-4f99d14d0c60.png)

OutPut: Can Still access admin panel from to different URL. 
![image](https://user-images.githubusercontent.com/90453683/169999261-6b20552f-43ae-42b3-bc87-1c532b4ca3e3.png)

**Reply by aris9ian** [reply]
I've noticed it yesterday too, I think it's done by purpose, because a lot could go wrong if the folder would get moved. It's better to copy it. What would happen if the "manager" folder already existed and some files inside had same filenames? The files would get overwritten etc etc

**Reply by condor2** [reply]
Of course.
Another fast release :))

**Reply by condor2** [reply]
Renaming admin folder is a stupid idea.
When admin folder should be the same and only URL should be changed: store.com/something (something =  redirect to admin page)

**Reply by IP-CAM** [reply]
With an (OC v.4.x) upgraded Extension like the one linked below, one could secure Admin Access very efficient, without renaming the Admin Directory:
OC v.3.x
https://www.opencart.com/index.php?route=marketplace/extension/info&extension_id=40693
OC v.2.x
https://www.opencart.com/index.php?route=marketplace/extension/info&extension_id=24045

**Comment by felipoantonoff** [comment]
Congratulations Daniel , team and all contributors, I was looking forward to Opencart 4 :)

I found it interesting to have fewer base extensions, so each user installs the desired one, I just really missed **OCMOD (XML)** I consider it the great and perhaps the best differential of Opencart 3 and 2 and basically every module is made using it, even if it's just to inform the version or other details.

Events is interesting, but I consider OCMOD much more powerful, with it I can change any Opencart file, be it controller, model, view or other without touching the original file and best of all, except file dependency, I can even assemble one controller without creating a file for it for example, it also facilitates maintenance and editing.

In the View itself is the great power of OCMOD, I can do it by installments in the products, let me know, change CSS, put JS in several desired places, in short, if there is a line to be researched to be done with OCMOD, since Event I believe that be much more limited.

I would recommend going back to OCMOD, I don't see the point in taking the same, even more being a great differential of Opencart and used by thousands of modules, much easier to keep 1 XML file than several PHP and Events files and I believe that Events have limitations mainly in the view , I don't know if it will be possible in any line to add something, remove something or change something.

It will also be very complex for most developers to maintain 2 totally different modules between Opencart 3 and 4, since in 3 most use OCMOD.

**Reply by condor2** [reply]
Now fix the bugs :))

**Reply by felipoantonoff** [reply]
I created a topic on this subject, it's an interesting debate and with some reasons to keep OCMOD in my opinion and that of other colleagues: https://github.com/opencart/opencart/discussions/10883

**Comment by cartbinder** [comment]
**Congratulations Daniel and the entire team for the OC4. Looking forward to that.**

A lot of great things were added in the new version.

The only thing that looks difficult currently is the event-based system.

**For adding stuff to themes or user interface:**
A lot of extensions need to add code at specific locations to work perfectly.
With event bases system it is necessary that the target element have some unique code at that specific location.
So it can be searched and replaced. Without that very difficult to add.
For this, **it will be necessary to have unique id/classes** for HTML elements on the list page ( order, customer, product ), and catalog pages ( front end ).

There are more problems related with the event-based system. Like the **stalker** said not able to modify the SQL query.
So to achieve the same thing via event, there would be multiple database calls. Because in each event, the developer will need to call and get the stuff he needs. 

Besides the above, if two extensions work on the same file. It will also be hard to make a patch. First, find that extension that is working on the same file. It is going to be a big pain for the developers.

**Ocmod system was/is  really great.** A lot of things could be achieved easily. Also easy to create patch files for working with other extensions.

**Comment by condor2** [comment]
Who tested the installer?
How ocmod.zip structure should be?
I have tested a language

For example if you use upload folder and install.json the path inserted in database will be

package_name/upload/extension/opencart/catalog/language/ro-ro/

And the language is not detected in admin

**Reply by aris9ian** [reply]
I did.

Structure of ocmod.zip should not follow the old way (upload dir). You just zip the admin, catalog directories and install.json file too.

As for the language mine is detected just fine using the following path:
public_html/extension/{extension_dir}/admin/language/en-gb/{extension type e.g. feed, module etc}/extension_filename.php

**Reply by stalker780** [reply]
Check ocmod.zip structure here
https://github.com/stalker780/opencart-vqmod-manager/releases

There should be no upload folder.

ocmod.zip can contain either single package:
| - admin/
| - catalog/
| - install.json

or multiple packages
my_extension_package.ocmod.zip:
|-my_extension_code1/
|-my_extension_code1/admin/
|-my_extension_code1/catalog/
|-my_extension_code1/system/
|-my_extension_code1/install.json
|-my_extension_code2/
|-my_extension_code2/admin/
|-my_extension_code2/catalog/
|-my_extension_code2/system/
|-my_extension_code2/install.json

**Reply by condor2** [reply]
I have changed but same issue with wrong path

something/extension/opencart/catalog/language/ro-ro/

something is the name of ocmod.zip file

when should be
opencart/extension/opencart/catalog/language/ro-ro/

**Reply by stalker780** [reply]
@condor2 
Are you trying to apply a language pack to opencart extensions?

**Reply by condor2** [reply]
Installing a language using the installer

**Reply by stalker780** [reply]
@condor2 
I am afraid you need to use events for this.
Copying files with installer as before is not enough

**Reply by condor2** [reply]
No need for event to install a language...
Work just fine if just copy paste.

Installer insert in database package name as path and not /opencart/

something/admin/language/ro-ro/localisation/address_format.php

When should be
opencart/admin/language/ro-ro/localisation/address_format.php

**Reply by stalker780** [reply]
That's a bad idea to copy language files to admin, catalog and extension folder.
You can do this manually, but you cannot do this with installer.

So, without using events you cannot create plug and play ocmod with installer.

Only events or vqmod can help you to rewrites language paths to your extension (language pack).
Just wait for events manual from Daniel.

This is a first example of poorly coded extension. Absence of ocmod should save from this, as osworx was talking about for so long :)

**Reply by condor2** [reply]
Using events only to upload some files through installer is a bad idea...

**Comment by karan982** [comment]
Does anybody know view file code like we do in Ocmod ?

for example this is view code:
              `{% for store in stores %}`
                `<a href="{{ store.href }}" target="_blank" class="dropdown-item">{{ store.name }}</a>`
              `{% endfor %}`
              
with ocmod we can set search string 
`<a href="{{ store.href }}" target="_blank" class="dropdown-item">{{ store.name }}</a>`
and replace with 
`<a href="{{ store.href }}" target="_blank" class="dropdown-item store-{{ store.id }}">{{ store.name }}</a>`

Download sample ocmod.zip in this link.
https://www.codinginspect.com/test/modify_viewfile.ocmod.zip

do not install this ocmod file on opencart store. it is only demonstration purpose only.  To acheive same out with event system.

**Does anybody know how we can acheive the same event system ?**

**Reply by felipoantonoff** [reply]
Hello, in this thread some colleagues gave some tips: https://github.com/opencart/opencart/discussions/10883

Here are some topics about:
For example: https://forum.opencart.com/viewtopic.php?f=144&t=227710
https://forum.opencart.com/viewtopic.php?f=24&t=211914
https://forum.opencart.com/viewtopic.php?t=221533

As I understand it, you need to pull the original content of the View to be modified, use str_replace or something like that to change it and then return the modified view, unfortunately it seems that it is not 100% prepared for this use, having to be done as needed the proper settings and code.

It seems that Events, at least in the View, can do close to what OCMOD could, but it seems much more complex and raw to me.

I'm curious how to move any file such as Controller, Model or Library, since View managed to understand how it does it, already moving any line of a PHP file I have no idea and if you want it to be possible, hopefully it will be.

**Reply by stalker780** [reply]
And don't forget to cache your tons of str_replaces and preg_matches :)
Otherwise they will be executed on every page request.

Also make this cache compatible with all other extensions events, which can parse view before or after you.

Especially it is funny to str_replace view after twig.

In OCMod you would replace `<div>{{ price }}</div>`

With events try to replace `<div>19.55EUR</div>` and `<div>9.99EUR</div>` when you have hundreds of not unique divs :)

**Reply by mhcwebdesign** [reply]
@stalker780: Wrong! Please read up on the view/*/before events on the OpenCart forum.

**Reply by stalker780** [reply]
@mhcwebdesign No problem. I will talk to you after you build a couple of complicated extensions with events only.
`*` mask is a perfect performance killer in OC4 events.

**Reply by felipoantonoff** [reply]
@stalker780 Nice summary, it doesn't make sense to use Events to replace OCMOD.

Now each module will have to take care of several resources, it will be a lot of work and in the end without any gain, each tool has its use, OCMOD is an excellent tool, even in a Podcast I talked about the great advantage of Opencart being completely modular and without having to touch 1 line in the original files, it seems like magic.

Back OCMOD (XML) ;)

**Comment by 0x00trust** [comment]
This version ready for live store or not?

**Reply by condor2** [reply]
Not yet

**Comment by condor2** [comment]
In menu, I think will be much ok if instead of Show all CategoryName to be Show all Products

**Comment by cocode32** [comment]
I've set up a new instance of open cart, but I cannot seem to turn on the SSL/HTTPS mode for the site. Please can someone help me?

I've seen in the docs that you do it in the admin portal, but that setting for SSL is no longer there in version 4.0.0.0
How can I enable my opencart site to run on HTTPS.

I've got the certificates and everything setup for the server hosting the files, but when I have it listen to 443, it does not load the page. It just times out

Any advice, please?

**Comment by cgcmart** [comment]
opencart did define https in config file.
It looks your instance needs turn on https on webserver config level.

**Reply by cocode32** [reply]
Ah, I see. I will give that a try! Thank you

**Reply by cocode32** [reply]
It worked, thank you!

**Comment by alecscodes** [comment]
Why did you make telephone optional and email not?
Many customers want the email to be optional, the telephone always need it for the courier.

Another thing, how can I add custom data for example show more custom information in order page in admin without ocmod? It's a mockery for developers!

**Reply by danielkerr** [reply]
body is asking for you advice nor should anybody!

**Reply by alecscodes** [reply]
It's an opinion, don't be sad.

**Comment by frameyourbag** [comment]
Hey guys,

Just an opinion from me, maybe on order info, Invoice and Dispatch Note add an column with a photo of the item?
Making it easier for the person who is preparing the order to be found?
Plus easier for the customer when is trying to find the item he ordered.
![order info](https://user-images.githubusercontent.com/114393100/192244506-8084d5d2-5365-4182-931c-2a07a03d5aa6.JPG)
![invoice example](https://user-images.githubusercontent.com/114393100/192244510-09d92e9f-5a1e-40a1-80a4-d017791a1a0f.JPG)

This is an example from Opencart 2.x.x.x
Thanks!

**Comment by Kushagrarenuka** [comment]
How can I make installable files for opencart version 4.x.x.x

**Comment by normdouglas** [comment]
Hello, @danielkerr  I'm a long time OC user... we have a large site and we have decided to move to OC4. However, we've already run into what are almost showstoppers... namely the methodology of SEO URLs. Before we start doing our own VQMOD to change the routing, is there a core way to remove the language? Also, is there new documentation because what I can find appears to just be OC3. We really want to make OC4 work, but I'm now wondering if it's ready for live sites?
Please this is not criticism, I've remained a loyal user of OC for a long time so just looking for some dialogue on this.
Cheers

**Reply by normdouglas** [reply]
Having recently chatted with @danielkerr I feel very confident that he and the project in general is moving in the right direction. I'm happy to delay our migration to OC4 and wait until the time is right.
Thank you to all involved... I've been involved in opensource projects in the past but I've never contributed to OC for this I feel guilty. But just know that the eComm industry respects OC and I can only imagine this will continue to grow in the coming months.
